Compilation fails when crossing midnight

I tried to install from scratch and the procedure lasted more than 24 hours on my Macbook AIR m4. However when crossing midnight it failed.

```
cd "/Users/pieter/Projects/m68k-atari-mint-cross-tools/compile/fdlibm-Git-20260216-bin-darwin/" && \
../fdlibm-Git-20260217/configure --host=m68k-atari-mint --target=m68k-atari-mint --prefix="/opt/cross-mint"

/bin/sh: line 1: ../fdlibm-Git-20260217/configure: No such file or directory
make[3]: *** [Makefile.fdlibm:87: /Users/pieter/Projects/m68k-atari-mint-cross-tools/compile/fdlibm-Git-20260216-bin-darwin/Makefile] Error 127
```
The reason is that the compilation took more than 24 hours, and the compilation of `fdlibm` had started on Feb. 16 just before midnight so the directory that was created was `fdlibm-Git-20260216-bin-darwin/` but the compilation continued until after midnight on Feb. 17 and then it was looking for `fdlibm-Git-20260217-bin-darwin/`.

The problem is caused by the definition of `VERSIONBUILD` in `VARS`
```
VERSIONBUILD=$(shell date +%Y%m%d)
```
This causes the shell command to be evaluated each time $(VERSIONBUILD) is used in the Makefile, instead of once at the beginning. This is just the way Makefile variables work.
This can be corrected by using
```
VERSIONBUILD:=$(shell date +%Y%m%d)
```
This will fix the value of `VERSIONBUILD` at the beginning of the make.

However, this is not enough.

After restarting the make, and crossing another day boundary, it failed with:
```
cat /Users/pieter/Projects/m68k-atari-mint-cross-tools/compile/gcc-Git-20260218/gcc/limitx.h /Users/pieter/Projects/m68k-atari-mint-cross-tools/compile/gcc-Git-20260218/gcc/glimits.h /Users/pieter/Projects/m68k-atari-mint-cross-tools/compile/gcc-Git-20260218/gcc/limity.h > /Users/pieter/Projects/m68k-atari-mint-cross-tools/compile/gcc-Git-20260218-bin-darwin-arm64/gcc/include-fixed/limits.h
/bin/sh: /Users/pieter/Projects/m68k-atari-mint-cross-tools/compile/gcc-Git-20260218-bin-darwin-arm64/gcc/include-fixed/limits.h: No such file or directory
make[1]: *** [Makefile.gcc-new:131: fix_limits] Error 1
```

`compile/gcc-Git-20260218` did not exist, instead there was `compile/gcc-Git-20260217`. This one was created by a different Makefile than the one that failed. So each one had its own value for `VERSIONBUILD`.

This problem must be solved by making sure all sub-makes use the same `VERSIONBUILD` value, and not each one its own value.
This can be done by replacing its definition in VARS by:
```
ifndef VERSIONBUILD
    VERSIONBUILD:=$(shell date +%Y%m%d)
endif
```
In this way, the main Makefile passes the value of `VERSIONBUILD` to all sub-makes.
Moreover, when you have to restart a make on a later day, you can pass the initial value like
```
VERSIONBUILD=20260216 make
make VERSIONBUILD=20260216 
```

I am sorry, but this setup with libcmini no longer makes it possible to use the original mintlib instead of libcmini. (I.e. compile without mentioning libcmini). The include files for mintlib and also crt0.o have been overwritten by those of libcmini.
So when someone needs the full mintlib features instead of libcmin, this might give problems.

I get an error on one of my programs if I compile without libcmini.: 
```
undefined reference to `__libc_ctype'.
```
So I think the libcmini stuff shoud be in a separate directory structure.

--- a/.github/workflows/make.yml
+++ b/.github/workflows/make.yml
@@ -22,10 +22,10 @@ jobs:
         run: make PREFIX=$RUNNER_TEMP/cross-mint all
 
       - name: Fast smoke tests
-        run: make -C tests PREFIX=$RUNNER_TEMP/cross-mint check
+        run: make PREFIX=$RUNNER_TEMP/cross-mint test
 
       - name: Full integration tests
-        run: make -C tests PREFIX=$RUNNER_TEMP/cross-mint check_all
+        run: make PREFIX=$RUNNER_TEMP/cross-mint test-all
 
       - name: Upload Artifact
         uses: actions/upload-artifact@v4

--- a/Makefile
+++ b/Makefile
@@ -8,7 +8,7 @@ ARCH2=i386
 ##########################################
 all:	init_dirs binutils mintbin gcc gemlib libcmini distrib
 
-.PHONY: check_autoconf check_libtool
+.PHONY: check_autoconf check_libtool test test-all test-qed
 
 check_autoconf:
 	@command -v autoconf >/dev/null 2>&1 || { \
@@ -46,6 +46,15 @@ gcc4:	check_libtool $(BUILD_DIR)/gcclibs gcc464
 
 binutils mintbin mintlib pml fdlibm gemlib libcmini cflib qed gcc464 gcc-new gemma:	init_dirs
 	$(MAKE) -f Makefile.$@
+
+test:
+	$(MAKE) -C tests PREFIX="$(PREFIX)" check
+
+test-all:
+	$(MAKE) -C tests PREFIX="$(PREFIX)" check_all
+
+test-qed:
+	$(MAKE) -C tests PREFIX="$(PREFIX)" qed_build_check
 
 distrib:
 	cd "$(dir $(PREFIX))" && \

--- a/README.md
+++ b/README.md
@@ -61,6 +61,36 @@ make
 
 This will build and install the tool in `/opt/cross-mint` and will produce a "distribution package" in the `packages` directory.
 
+## Tests
+
+Basic test programs and checks are available in the `tests/` folder.
+
+Run fast smoke tests (hello world default/GEM/minimal):
+
+```bash
+make test
+```
+
+Run full integration tests (includes building and checking QED):
+
+```bash
+make test-all
+```
+
+Run only the QED build and checks:
+
+```bash
+make test-qed
+```
+
+When using a custom toolchain location (e.g. in CI), pass `PREFIX`:
+
+```bash
+make PREFIX=$RUNNER_TEMP/cross-mint all
+make PREFIX=$RUNNER_TEMP/cross-mint test
+make PREFIX=$RUNNER_TEMP/cross-mint test-all
+```
+
 ### Known problems
 
 On older macOS versions (e.g. macOS Sierra 10.13) where the `tar` command is unable to extract `bz2` or `lz` compressed files, you will encounter the error message `tar: Unrecognized archive format` while `gcclibs` is being executed.

--- a/VARS
+++ b/VARS
@@ -25,5 +25,12 @@ HOMEBREW_PREFIX=$(shell brew config | grep HOMEBREW_PREFIX | awk '{print $$2}')
 ##########################################
 VERSIONBIN=bin-$(HOSTSYS)
 VERSIONBINCPU=$(VERSIONBIN)-$(ARCH)
-VERSIONBUILD ?= $(shell date +%Y%m%d)
-VERSIONBUILD := $(VERSIONBUILD)
+VERSIONBUILD_FILE ?= $(BUILD_DIR)/.versionbuild
+VERSIONBUILD ?= $(shell mkdir -p "$(BUILD_DIR)"; \
+	if [ -f "$(VERSIONBUILD_FILE)" ]; then \
+		cat "$(VERSIONBUILD_FILE)"; \
+	else \
+		date +%Y%m%d > "$(VERSIONBUILD_FILE)"; \
+		cat "$(VERSIONBUILD_FILE)"; \
+	fi)
+VERSIONBUILD := $(strip $(VERSIONBUILD))

--- a/tests/Makefile
+++ b/tests/Makefile
@@ -5,7 +5,7 @@ CC = $(PREFIX)/bin/$(TARGET)-gcc
 OBJDUMP = $(PREFIX)/bin/$(TARGET)-objdump
 FILE_CMD ?= file
 
-BUILD_DIR ?= build
+TEST_BUILD_DIR ?= build
 LIBCMINI_ROOT ?= $(PREFIX)/m68k-atari-mint/libcmini
 ROOT_DIR ?= ..
 
@@ -17,10 +17,10 @@ LIBCMINI_CFLAGS ?= -I$(LIBCMINI_ROOT)/include
 LIBCMINI_LDFLAGS ?= -s -nostdlib $(LIBCMINI_ROOT)/lib/crt0.o
 LIBCMINI_LDLIBS ?= -L$(LIBCMINI_ROOT)/lib -lgcc -lcmini -lgcc
 
-HELLO_DEFAULT_BIN := $(BUILD_DIR)/hello-default.tos
-HELLO_MINIMAL_OBJ := $(BUILD_DIR)/hello-minimal.o
-HELLO_MINIMAL_BIN := $(BUILD_DIR)/hello-minimal.tos
-HELLO_GEM_BIN := $(BUILD_DIR)/hello-gem.prg
+HELLO_DEFAULT_BIN := $(TEST_BUILD_DIR)/hello-default.tos
+HELLO_MINIMAL_OBJ := $(TEST_BUILD_DIR)/hello-minimal.o
+HELLO_MINIMAL_BIN := $(TEST_BUILD_DIR)/hello-minimal.tos
+HELLO_GEM_BIN := $(TEST_BUILD_DIR)/hello-gem.prg
 
 .PHONY: all clean check check_all \
 	hello_world hello_world_gem hello_world_minimal \
@@ -53,22 +53,22 @@ qed_check:
 qed_build_check: qed_build qed_check
 	@echo "QED build and validation passed."
 
-$(BUILD_DIR):
+$(TEST_BUILD_DIR):
 	mkdir -p "$@"
 
-$(HELLO_DEFAULT_BIN): hello.c | $(BUILD_DIR)
+$(HELLO_DEFAULT_BIN): hello.c | $(TEST_BUILD_DIR)
 	$(CC) $(CFLAGS_COMMON) $(LDFLAGS_COMMON) -o "$@" "$<" $(LDLIBS_COMMON)
 	@$(FILE_CMD) "$@" | grep -q "Atari ST M68K contiguous executable"
 	@$(OBJDUMP) -f "$@" | grep -q "file format a.out-mintprg"
 	@$(OBJDUMP) -t "$@" | grep -q "___libc_main"
 
-$(HELLO_GEM_BIN): hello_gem.c | $(BUILD_DIR)
+$(HELLO_GEM_BIN): hello_gem.c | $(TEST_BUILD_DIR)
 	$(CC) $(CFLAGS_COMMON) -o "$@" "$<" -lgem
 	@$(FILE_CMD) "$@" | grep -q "Atari ST M68K contiguous executable"
 	@$(OBJDUMP) -f "$@" | grep -q "file format a.out-mintprg"
 	@$(OBJDUMP) -t "$@" | grep -q "_mt_form_alert"
 
-$(HELLO_MINIMAL_OBJ): hello.c | $(BUILD_DIR)
+$(HELLO_MINIMAL_OBJ): hello.c | $(TEST_BUILD_DIR)
 	$(CC) $(CFLAGS_COMMON) $(LIBCMINI_CFLAGS) -c -o "$@" "$<"
 
 $(HELLO_MINIMAL_BIN): $(HELLO_MINIMAL_OBJ)
@@ -79,4 +79,4 @@ $(HELLO_MINIMAL_BIN): $(HELLO_MINIMAL_OBJ)
 	@test "$$(wc -c < "$@")" -lt "$$(wc -c < "$(HELLO_DEFAULT_BIN)")"
 
 clean:
-	rm -rf "$(BUILD_DIR)"
+	rm -rf "$(TEST_BUILD_DIR)"

